### PR TITLE
vargadigiacomo/tirocinio-plugin-sns

### PR DESCRIPTION
**Pull request by Zhin1797**
- createTopic
- Publish

### CODE REVIEW

File: packages/plugins/sns/src/aws-sns/index.ts [hunk 1]
```diff
@@ -0,0 +1,74 @@
+import { SNS, Credentials } from 'aws-sdk';
+
+class AwsSNS {
+  awsSNS: SNS
+
+  init() {
+    const SNSConfigOptions:SNS.ClientConfiguration = {
+      credentials: new Credentials(process.env['RLN_SES_AWS_ACCESS_KEY_ID'], process.env['RLN_SES_AWS_SECRET_ACCESS_KEY']),
+      region: process.env['RLN_SES_AWS_REGION'],
```
Comment (antoniogiordano) [inline packages/plugins/sns/src/aws-sns/index.ts:9]:
```suggestion
      credentials: new Credentials(process.env['RLN_SNS_AWS_ACCESS_KEY_ID'], process.env['RLN_SNS_AWS_SECRET_ACCESS_KEY']),
      region: process.env['RLN_SNS_AWS_REGION'],
```

File: packages/plugins/sns/src/aws-sns/index.ts [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+    const params: SNS.Types.CreateTopicInput = {
+      Name: topicName,
+      Attributes:{
+        DisplayName: displayName,
+      },
+    };
+    return await this.awsSNS.createTopic(params).promise();
+  }
+
+  async publish(message: string, ARN: string, messageTitle: string){
```
Comment (antoniogiordano) [inline packages/plugins/sns/src/aws-sns/index.ts:55]:
```suggestion
  async publish(title: string, ARN: string, message: string){
```

File: packages/plugins/sns/src/aws-sns/index.ts [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+    const params: SNS.Types.PublishInput = {
+      MessageStructure: 'JSON',
+      Message: JSON.stringify({
+        default: message,
+        title: messageTitle,
+      }),
+      TopicArn: ARN,
+    };
```
Comment (antoniogiordano) [inline packages/plugins/sns/src/aws-sns/index.ts:63]: Here we should change for sure.
This is the template I'm using right now:
```{
  "default": "Nuova Promozione",
  "APNS": "{\"aps\":{\"alert\": {\"title\": \"MerchantName\", \"body\": \"Nuova Promozione!\"}, \"sound\": \"bingbong.aiff\"}, \"eventId\": \"6044aa87ac6850ac512096af\" }",
  "GCM": "{ \"data\": { \"title\": \"MerchantName\", \"body\": \"Nuova Promozione!\", \"vibrate\": true, \"eventId\": \"6044aa87ac6850ac512096af\" } }"
}
```
So I think we should specify we are using a custom template to AWS

**Review by antoniogiordano — CHANGES_REQUESTED** [review]

File: packages/plugins/sns/src/aws-sns/index.ts (hunk 3 above)
Comment (Zhin1797) [reply packages/plugins/sns/src/aws-sns/index.ts]: For the custom template we are using MessageStructure: 'json' and we are defining the message with JSON.stringify, we added APNS and GCM objects but we don't know if that's the correct approach.

File: packages/plugins/sns/src/aws-sns/index.ts [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+            body: message,
+            vibrate: true,
+            eventId: eventId,
+          },
+        },
+      }),
+      TopicArn: ARN,
+    };
+    return await this.awsSNS.publish(params).promise();
+  }
```
Comment (antoniogiordano) [inline packages/plugins/sns/src/aws-sns/index.ts:82]: Remember this is a generic plugin, not just for Adhesion :-) so we can not make it too much strict.
It was better as it was before sorry :-) let's decide the project that installs the plugin which payload should be sent!
But this code will be useful project side

Comment (Zhin1797) [reply packages/plugins/sns/src/aws-sns/index.ts:82]: oh yes, sorry, but that's the correct way to handle custom messages? Sice we don't want to messup AWS with tests we wanted to ask you if that's the correct way to handle SNS publishings.

Comment (antoniogiordano) [reply packages/plugins/sns/src/aws-sns/index.ts:82]: I have no idea, I sent them with the AWS dashboard :D we will know only when we'll try ;-) don't worry!

**Review by antoniogiordano — CHANGES_REQUESTED** [review]
Great, let's add some other APIs from old project, and we are done!

**Review by antoniogiordano — APPROVED** [review]

**Pull request merged**